### ISSUE TITLE
What would be needed to implement interaction values for other Explainers? For which one would it be most straightforward?

### ISSUE THREAD

**Issue by MichalPorwisz** [shap/shap#777]
I (and a few other people) am interested in using interaction values for models different than tree ensembles and/or for tree ensembles, but with different backgrounds than just training set (so with feature_dependence="independent"). If that wouldn't be that difficult and you would provide some guidance, we could consider taking a go at implementing them ourselves.

As stated in the paper, SHAP interaction values can be interpreted as the difference between the SHAP values for feature _i_ when feature _j_ is present and the SHAP values for feature _i_ when feature _j_ is absent. Also, Shapley values equation requires that for obtaining SHAP value for feature _i_, one has to calculate contributions of all possible coalitions, so this includes all those when feature _j_ is present and those where it is absent. This would mean that we just need to keep track of those coalitions and calculate separate average of those with _j_ present and those with _j_ absent (and do this for all other features). Is this correct or am I missing something? How do the approximations used by different explainers impact that idea?

Another way would be to run an explainer twice for each feature while one input is constrained to either be the current input or the background, then taking the difference, as you mentioned here:
https://github.com/slundberg/shap/issues/272#issuecomment-425678211
That would be especially straightforward, if an Explainer uses internally some function, where you can set which feature is taken from current input and which is integrated out. Also, you mention possible issues due to some approximations in DeepExplainer, but maybe for other explainers there wouldn't be such problems?

Additionally, in which of all the explainers (other than TreeSHAP with feature_dependence: "tree_path_dependent") would you say that implementing interaction values should be most straightforward?

**Comment by slundberg** [comment]
Great questions!

- Getting it to work for TreeExplainer with `feature_dependence="independent"` would be the easiest, but would still take a very non-trivial modification of the C++ code and a careful consideration of the interaction effect definition.
- Doing it for the Deep and Gradient explainers would require some careful consideration. You could just try what I mentioned in #272, and this might be easier even than the above option if it proves simple.   
- Doing it for Kernel or Partition explainer would be doable, but how practical it would be remains to be seen (since computing them is much slower than just the original values).

If you have specific ideas you want to run by me, please do. And I am also happy to help if you want to tackle any of them (though it will be a few weeks before I have much time to sink into it). Hope that's helpful!
